### ISSUE TITLE
Allow access to custom pytree definitions and override them

### ISSUE THREAD

**Issue by inversecrime** [google/jax#21666]
Currently, you can do this:
```python
from typing import Any, Callable, TypeVar

import jax._src.tree_util as _jtu

C = TypeVar("C")


def get_registered_tree_functions(cls: type[C]) -> tuple[Callable[[C], tuple[Any, Any]], Callable[[Any, Any], C]]:
    flatten = _jtu._registry[cls].to_iter  # type: ignore
    unflatten = _jtu._registry[cls].from_iter  # type: ignore
    return (flatten, unflatten)
```
which is not public API and doesn't cover all cases.

It would be nice to have access to this via some public API, and it would be nice to modify already registered nodes.
This could be extremely useful for i.e. class decorators.

Currently, if you wanted to modify the pytree-behavior of a class, you would have to replace the class with a copy of that class, assign a new name to that new class, and register it under the new name.
... Maybe there is a more straightforward approach, but I cannot think of one.

**Comment by jakevdp** [comment]
No, this is a private API, and for good reason! `jax._src.tree_util._registry` is not the only pytree registry that comes into play, and so your example code would fail to overrid the dispatch-path registrations. We could maybe provide an API to override existing registrations, but that would be problematic for other reasons: for example, there are cacheing mechanisms in JAX's dispatch path that depend on pytree registrations being static, so making that change would be fairly intrusive.

I think a better approach is to register a new class if you want new flattening/unflattening behavior.

**Comment by inversecrime** [comment]
Thanks for the reply!

I guess in the case of class decorators, I could just use my custom decorator _before_ `register_pytree_node_class`, and let the custom decorator change the methods `tree_flatten` and `tree_unflatten` of that class before it gets even registered.

In other cases however, it would be nice to have access to a node's `flatten` and `unflatten` method - even if overriding them is not possible.

**Comment by inversecrime** [comment]
It should be possible to access the children of a pytree by calling `jax.tree.flatten` with a specific `is_leaf` argument.
But you cannot get auxiliary data this way.

**Comment by jakevdp** [comment]
Can you say more about your larger goal here? I'm not sure I fully understand the context of your question.

**Comment by inversecrime** [comment]
Hm, that's quite complicated.

The number one feature I would wish for is what I call "Tree Arrays". To my knowledge, there is currently no library that implements this.

In my opinion, one of the best features of jax is that you can use all transformations (especially `vmap`) with pytrees.
Now, when you use a `vmap` which outputs a pytree, the pytree is can be seen as a "pytree array" - all it's leaves are mapped over the same axis (unless otherwise specified).

An pseudo-example could be this:
```python
@tree_array
@tree_dataclass
class Sphere:
    center: Array
    radius: Array

# Create a (7, 4)-array of 3d spheres
spheres = jax.vmap(jax.vmap(
    lambda center: Sphere(center=center, radius=jnp.asarray(1.0))
))(jax.random.normal(key, shape=(7, 4, 3)))

print(spheres.shape)  # Returns (7, 4)
print(spheres[3, 0])  # Returns the sphere at position (0, 3)
print(spheres.at[1, 2].set(0))  # Replaces the sphere at position (1, 2) with center and radius all zero
```

Here, `tree_dataclass` can be any decorator that makes the class a jax-compatible dataclass.
The decorator `tree_array` then makes the class behave like an array.

I have already implemented this and use it a lot. The main part of the `tree_array` decorator is this:
```python
cls = type(cls.__name__, (cls, TreeArrayImpl), cls.__dict__)
```
This basically "copies" the class and adds a new base to it which implements all the functionality. This new class can then be registered as a pytree. So far it works fine, maybe a bit hacky, but nothing really crazy.

This approach does **not** require overriding pytree definitions.

The problem is now that I need access to the original classes `flatten` and `unflatten` methods, so that I can modify them (i.e. add a `shape` property to the auxiliary data).

Currently, I am using the `get_registered_tree_functions` function I posted originally to do this, but it would be nice to have a better way to access them.

**Comment by jakevdp** [comment]
Thanks - you may be able to do this without having to directly access the flattening rule. Here's an example of what it might look like:
```python
import jax
from dataclasses import dataclass
from typing import Any

@dataclass
class MyTree:
  x: jax.Array
  y: str
jax.tree_util.register_dataclass(MyTree, data_fields=['x'], meta_fields=['y'])


@jax.tree_util.register_pytree_node_class
class Derived(MyTree):
  shape: tuple[int, ...]

  def __init__(self, x, y, shape):
    self.x = x
    self.y = y
    self.shape = shape

  def tree_flatten(self):
    # Construct a MyTree object from a Derived object
    # There may be a better way to do this in your application.
    base = object.__new__(MyTree)
    base.__dict__ = self.__dict__
    return [base], self.shape

  @classmethod
  def tree_unflatten(cls, aux_data, children):
    # Construct a Derived object from a MyTree object
    # There may be a better way to do this in your application.
    base, = children
    obj = object.__new__(Derived)
    obj.__dict__ = base.__dict__
    obj.shape = aux_data
    return obj

d = Derived(x=jax.numpy.arange(4), y='abc', shape=(1, 2))
print(d)
# Derived(x=Array([0, 1, 2, 3], dtype=int32), y='abc')
print(d.shape)
# (1, 2)

leaves, tree = jax.tree.flatten(d)
print(leaves)
# [Array([0, 1, 2, 3], dtype=int32)]

d2 = jax.tree.unflatten(tree, leaves)
print(d2)
# Derived(x=Array([0, 1, 2, 3], dtype=int32), y='abc')
print(d2.shape)
# (1, 2)
```
The main thing this relies on is the ability to construct a `Base` object from a `Derived` object. Python doesn't offer any general way to do this as far as I can tell, but you can do so through various means in specific situations. Hope that helps!

**Comment by inversecrime** [comment]
This is a great suggestion and exactly what I wanted.

Using this approach, you can arbitrarily extend the pytree definition of a given class without knowing how that class looks like - perfect for a decorator.

Thank you!
